File: frontend/src/services/study.ts
import { api } from "./api";

export interface DueCard {
  id: number;
  front_content: string;
  back_content: string;
  deck_id: number;
  deck_title: string;
  ease_factor: number;
  interval: number;
  repetitions: number;
  next_review: string;
}

export interface ReviewRequest {
  card_id: number;
  quality: number; // 0-5
  rating?: string; // "again" | "hard" | "good" | "easy"
}

export interface ReviewResponse {
  card_id: number;
  new_ease_factor: number;
  new_interval: number;
  new_repetitions: number;
  next_review: string;
  xp_earned: number;
  coins_earned: number;
  new_streak: number;
}

export interface StudySessionResponse {
  total_cards: number;
  cards_reviewed: number;
  total_xp_earned: number;
  total_coins_earned: number;
  new_streak: number;
  session_duration_seconds: number;
}

export const studyService = {
  async getDueCards(deckId?: number, limit: number = 50): Promise<DueCard[]> {
    const params: Record<string, any> = { limit };
    if (deckId) {
      params.deck_id = deckId;
    }
    const response = await api.get<DueCard[]>("/study/due", { params });
    return response.data;
  },

  async submitReview(request: ReviewRequest): Promise<ReviewResponse> {
    const response = await api.post<ReviewResponse>("/study/review", request);
    return response.data;
  },

  async completeSession(data: {
    total_cards: number;
    cards_reviewed: number;
    total_xp_earned: number;
    total_coins_earned: number;
    session_duration_seconds: number;
  }): Promise<StudySessionResponse> {
    const response = await api.post<StudySessionResponse>("/study/session/complete", data);
    return response.data;
  },
};

File: backend/app/schemas/study.py
from datetime import datetime

from pydantic import BaseModel, Field


class ReviewRequest(BaseModel):
    """Request schema for submitting a card review."""

    card_id: int = Field(..., description="ID of the card being reviewed")
    quality: int = Field(..., ge=0, le=5, description="Quality rating (0-5)")
    rating: str | None = Field(
        None,
        description="Simple rating (again/hard/good/easy). If provided, overrides quality.",
    )


class ReviewResponse(BaseModel):
    """Response schema for a card review."""

    card_id: int
    new_ease_factor: float
    new_interval: int
    new_repetitions: int
    next_review: datetime
    xp_earned: int
    coins_earned: int
    new_streak: int


class StudySessionResponse(BaseModel):
    """Response schema for study session summary."""

    total_cards: int
    cards_reviewed: int
    total_xp_earned: int
    total_coins_earned: int
    new_streak: int
    session_duration_seconds: int


class DueCardResponse(BaseModel):
    """Response schema for a card due for review."""

    id: int
    front_content: str
    back_content: str
    deck_id: int
    deck_title: str
    ease_factor: float
    interval: int
    repetitions: int
    next_review: datetime

class SessionCompleteRequest(BaseModel):
    """Request schema for completing a study session."""
    
    total_cards: int = Field(..., ge=0, description="Total cards in the session")
    cards_reviewed: int = Field(..., ge=0, description="Number of cards reviewed")
    total_xp_earned: int = Field(..., ge=0, description="Total XP earned in session")
    total_coins_earned: int = Field(..., ge=0, description="Total coins earned in session")
    session_duration_seconds: int = Field(..., ge=0, description="Duration of session in seconds")